pkg/tessera/tessera.go

```diff
@@ -0,0 +1,139 @@
+}
+
+// NewStorage creates a Tessera storage object for the provided driver and signer.
+func NewStorage(ctx context.Context, origin string, driver tessera.Driver, signer signature.Signer) (*Storage, error) {
+	noteSigner, err := note.NewNoteSigner(ctx, origin, signer)
+	if err != nil {
+		return nil, fmt.Errorf("getting note signer: %w", err)
+	}
+	appender, reader, err := tessera.NewAppender(ctx, driver,
+		tessera.NewAppendOptions().WithCheckpointSigner(noteSigner),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("getting tessera appender: %w", err)
+	}
+	awaiter := tessera.NewIntegrationAwaiter(ctx, reader.ReadCheckpoint, 1*time.Second)
+	return &Storage{
+		origin:     origin,
+		awaiter:    awaiter,
+		addFn:      appender.Add,
+		readTileFn: reader.ReadTile,
+	}, nil
+}
+
+// Add adds a Tessera entry to the log, waits for it to be sequenced into the log,
+// and returns the log index and inclusion proof as a TransparencyLogEntry object.
+func (s *Storage) Add(ctx context.Context, entry *tessera.Entry) (*rekor_pb.TransparencyLogEntry, error) {
+	idx, checkpointBody, err := s.addEntry(ctx, entry)
+	if err != nil {
+		return nil, fmt.Errorf("add entry: %w", err)
+	}
+	inclusionProof, err := s.buildProof(ctx, idx, checkpointBody)
+	if err != nil {
+		return nil, fmt.Errorf("building inclusion proof: %w", err)
+	}
+	return &rekor_pb.TransparencyLogEntry{
+		LogIndex:       idx.i,
+		InclusionProof: inclusionProof,
+	}, nil
+}
+
+// ReadTile looks up the tile at the given level, index within the level, and
+// width of the tile if partial, and returns the raw bytes of the tile.
+func (s *Storage) ReadTile(ctx context.Context, level, index uint64, p uint8) ([]byte, error) {
+	tile, err := s.readTileFn(ctx, level, index, p)
+	if err != nil {
+		return nil, fmt.Errorf("reading tile level %d index %d p %d: %w", level, index, p, err)
+	}
+	return tile, nil
+}
+
+func (s *Storage) addEntry(ctx context.Context, entry *tessera.Entry) (*safeInt64, []byte, error) {
+	idx, checkpointBody, err := s.awaiter.Await(ctx, s.addFn(ctx, entry))
+	if err != nil {
+		return nil, nil, fmt.Errorf("await: %w", err)
+	}
+	safeIdx, err := newSafeInt64(idx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid index: %w", err)
+	}
+	return safeIdx, checkpointBody, nil
+}
+
+func (s *Storage) buildProof(ctx context.Context, idx *safeInt64, signedCheckpoint []byte) (*rekor_pb.InclusionProof, error) {
+	checkpoint, err := unmarshalCheckpoint(signedCheckpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling checkpoint: %w", err)
+	}
+	proofBuilder, err := client.NewProofBuilder(ctx, checkpoint, s.ReadTile)
+	if err != nil {
+		return nil, fmt.Errorf("new proof builder: %w", err)
+	}
+	proof, err := proofBuilder.InclusionProof(ctx, idx.u)
+	if err != nil {
+		return nil, fmt.Errorf("generating inclusion proof: %w", err)
+	}
+	hashes := make([][]byte, len(proof))
+	for i, h := range proof {
+		dst := make([]byte, hex.EncodedLen(len(h)))
```

Comment:
I don’t think the hashes should be hex-encoded. 

The Rekor verifier expects hex encoded hashes https://github.com/sigstore/rekor/blob/26bfbf550e1c3b086851d0ccbf2d8ebf024c7f46/pkg/verify/verify.go#L141, but Sigstore-go will handle that - https://github.com/sigstore/sigstore-go/blob/797ca59e9032c12c7e4a39092eb91c9d7c1b3b63/pkg/tlog/entry.go#L118

We can write a new verifier library later on that doesn’t rely on the generated OpenAPI models.

Reply:
Encoding bytes should be handled by the JSON marshaling library, as [base64](https://pkg.go.dev/encoding/json#Marshal:~:text=except%20that%20%5B%5Dbyte%20encodes%20as%20a%20base64%2Dencoded%20string).

IIRC Rekor v1 [hex-encoded bytes](https://github.com/sigstore/rekor/blob/26bfbf550e1c3b086851d0ccbf2d8ebf024c7f46/openapi.yaml#L632) to turn it into a string because OpenAPI 2 didn't properly handle bytes as return types.